### ISSUE TITLE
sim experiment - excitation wavelenght on dv extended header

### ISSUE THREAD

**Issue by carandraug** [MicronOxford/cockpit#290]
The base header of a mrc file has the emission wavelength. For SIM reconstruction we need the excitation wavelength. The mrc format does not have a place for this. The dv format seems to store it on the extended header but we don't have the specs for that.

This issue is point 2 on issue #288 

We need to ask GE for the extended header specs and/or see what bioformats does. Seems like the OME team have some specs for the extended header but are unable to share them.  We should be able to guess from their [source code](https://github.com/openmicroscopy/bioformats/blob/develop/components/formats-gpl/src/loci/formats/in/DeltavisionReader.java) . See also bioformats page on the [dv format](https://docs.openmicroscopy.org/bio-formats/5.7.2/formats/deltavision.html) and a recent [bug report about wanting more up to date specs](https://github.com/openmicroscopy/bioformats/issues/2656#issuecomment-314662950)

**Comment by carandraug** [comment]
I have started to fix this with commit carandraug/cockpit@a20c1a847ac70dd931253b8a01e68c0352e69b0a

I need someone's help to actually get the required metadata which I don't know where is kept. See the three TODO's on that commit.

Where is the excitation wavelength for the each image? The `CameraHandler.wavelength` property is the emission not excitation, correct?

Also, I guess we could add the exposure time. There is `CameraHandler.getExposureTime` but if the writing of each plane is not synced with the image acquisition, then it's possible that at the time of writing, the camera exposure settings are different from the actual settings used for acquisition, correct?

**Comment by carandraug** [comment]
A side note because the only reason to have all this trouble is to use softWoRx for reconstruction.

Is the excitation wavelength information in the extended header used by softworx to create a correct call to `OMXSIReconstruction`? Or is the extended header inspected by `OMXSIReconstruction` itself? Because it seems that the excitation wavelength is a parameter to call the reconstruction program, like so:

    OMXS [...] -channel_ex_linespacing $EX_WAVELENGTH $LINESPACING [...]

and so we could use it directly and not via softWoRx until we replace softWoRx.

**Comment by carandraug** [comment]
I have pushed commits 26e32f7bd9427d79ad7bee6e087bc9095b17a7d7...6ef282b9279911250cc7223408deea675c9fd0f1 after testing this on DeepSIM with Ian. This writes an extended header of 14 floats per image plane that bioformats and the ImageJ Deltavision Opener plugin can read. However, `OMXSIReconstruction` still fails to pick the excitation wavelength values. We don't know where `OMXSIReconstruction` is looking for that data.

The extended header is a group of floats and ints per image plane. The existing open source readers skip all the ints and read the first 14 floats. So the changes pushed now write only 14 floats. But reading actual dv files, it seems that it's writing 8 ints and 32 floats. We have no idea what may go on the other values, but maybe that shape of extended header is required to make `OMXSIReconstruction` happy:

```
>>> import struct
>>> fh = open(..., 'r')
>>> fh.seek(128)
>>> struct.unpack('h', fh.read(2)) # number of int
(8,)
>>> struct.unpack('h', fh.read(2)) # number of float
(32,)
>>> fh.seek(1024)
>>> struct.unpack('8i32f', fh.read(8*4 + 32*4))
(0, 0, 0, 0, 0, 0, 0, 0, 0.0, 0.0, 16640.0, 13462.0, 5520.75, -4002.38134765625, 4026.86279296875, 125.80743408203125, 60.0, 1.0, 488.0, 525.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
```

**Comment by carandraug** [comment]
Writing down 8 ints and 32 floats per plane, even though all filled with zero fixed the issue, i.e, softworx is now picking the right excitation wavelength for reconstruction. Pushed fix in commit 961a3cc. Closing as fixed.

Note that we don't know what all that new data is meant to have on the extended header. I left the code in place ready to fill the metadata if we ever get to know about them.

**Comment by iandobbie** [comment]
From the list I would take 488 as the excitation wavelength and 525 as the emission

Ian



Sent from my iPhone

On 23 Jul 2018, at 19:43, Carnë Draug <notifications@github.com<mailto:notifications@github.com>> wrote:


Closed #290<https://github.com/MicronOxford/cockpit/issues/290>.
